vector/src/main/java/im/vector/app/features/voice/AbstractVoiceRecorder.kt

```diff
@@ -57,9 +61,22 @@ abstract class AbstractVoiceRecorder(
         }
     }
 
-    override fun startRecord() {
+    override fun initializeRecord(roomId: String, attachmentData: ContentAttachmentData) {
+        getFilenameFromUri(context, attachmentData.queryUri)?.let {
+            val voiceMessageFolder = File(outputDirectory, roomId.md5())
+            outputFile = File(voiceMessageFolder, it)
+        }
+    }
+
+    override fun startRecord(roomId: String) {
         init()
-        outputFile = File(outputDirectory, "Voice message.$filenameExt")
+        val fileName = "Voice message-${UUID.randomUUID()}.$filenameExt"
```

Comment:
using unique files for the recording fixes #3833

Reply:
The filename can in some circumstance be visible to the user, is it possible to use a subfolder instead, and keep a human readable filename?

Reply:
<img width="293" alt="image" src="https://user-images.githubusercontent.com/3940906/142905311-ca32e2ad-3e16-41d9-b1d2-f29cb6323013.png">

Reply:
I'll have a look, out of interest do we need to display the real file name here, could the label be hardcoded?

Reply:
Yes, this is another way to fix the display issue. But also the file can be downloaded, etc. so a human readable filename should still be nice.

Reply:
https://github.com/vector-im/element-android/pull/4527/commits/c2ec15726cb52fd85f3bbaf08c8c2d39f9752965

We can use the file provider to provide a human readable name, downloads automatically append a timestamp

| NAME | DOWNLOAD | 
| --- | --- |
|![after-voice-message](https://user-images.githubusercontent.com/1848238/143029536-564e99c2-a0ee-4a42-af28-44b259da0523.png)|![after-downloaded-vm](https://user-images.githubusercontent.com/1848238/143029532-759e2045-3f48-434b-8b7e-9abb49bc1996.png)